rpc/client.go

```diff
@@ -0,0 +1,737 @@
+	switch resp, err := op.wait(ctx); {
+	case err != nil:
+		return err
+	case resp.Error != nil:
+		return resp.Error
+	case len(resp.Result) == 0:
+		return ErrNoResult
+	default:
+		return json.Unmarshal(resp.Result, &result)
+	}
+}
+
+// BatchCall sends all given requests as a single batch
+// and waits for the server to return a response for all of them.
+//
+// In contrast to Call, BatchCall only returns I/O errors.
+// Any error specific to a request is reported through the Error field
+// of the corresponding BatchElem.
+func (c *Client) BatchCall(b []BatchElem) error {
+	ctx := context.Background()
+	return c.BatchCallContext(ctx, b)
+}
+
+// BatchCall sends all given requests as a single batch
+// and waits for the server to return a response for all of them.
+// The wait duration is bounded by the context's deadline.
+//
+// In contrast to CallContext, BatchCallContext only returns I/O errors.
+// Any error specific to a request is reported through the Error field
+// of the corresponding BatchElem.
+func (c *Client) BatchCallContext(ctx context.Context, b []BatchElem) error {
+	msgs := make([]*jsonrpcMessage, len(b))
+	op := &requestOp{
+		ids:  make([]json.RawMessage, len(b)),
+		resp: make(chan *jsonrpcMessage, len(b)),
+	}
+	for i, elem := range b {
+		msg, err := c.newMessage(elem.Method, elem.Args...)
+		if err != nil {
+			return err
+		}
+		msgs[i] = msg
+		op.ids[i] = msg.ID
+	}
+
+	var err error
+	if _, ok := c.writeConn.(*httpConn); ok {
+		err = c.sendBatchHTTP(ctx, op, msgs)
+	} else {
+		err = c.send(ctx, op, msgs)
+	}
+
+	// Wait for all responses to come back.
+	for n := 0; n < len(b) && err == nil; n++ {
+		var resp *jsonrpcMessage
+		resp, err = op.wait(ctx)
+		if err != nil {
+			break
+		}
+		// Find the element corresponding to this response.
+		// The element is guaranteed to be present because dispatch
+		// only sends valid IDs to our channel.
+		var elem *BatchElem
+		for i := range msgs {
+			if bytes.Equal(msgs[i].ID, resp.ID) {
+				elem = &b[i]
+				break
+			}
+		}
+		if resp.Error != nil {
+			elem.Error = resp.Error
+			continue
+		}
+		if len(resp.Result) == 0 {
+			elem.Error = ErrNoResult
+			continue
+		}
+		elem.Error = json.Unmarshal(resp.Result, elem.Result)
+	}
+	return err
+}
+
+// EthSubscribe calls the "eth_subscribe" method with the given arguments, registering a
+// subscription. Server notifications for the subscription are sent to the given channel.
+// The element type of the channel must match the expected type of content returned by the
+// subscription.
+
+// Callers should not use the same channel for multiple calls to EthSubscribe. The channel
+// is closed when the notification is unsubscribed or an error occurs. The error can be
+// retrieved via the Err method of the subscription.
+//
+// Slow subscribers will block the clients ingress path eventually.
+func (c *Client) EthSubscribe(channel interface{}, args ...interface{}) (*ClientSubscription, error) {
```

Comment:
Not sure if this is the right direction to go. In the future we probably have subscriptions for whisper and swarm as well. These will be prefix with a different identifier. Maybe this method can accept the namespace as an argument?